jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java

```diff
@@ -65,78 +65,238 @@ String getName() {
+      addLayerFile(LayerType.RESOURCES, sourceFile, pathInContainer, false);
+      return this;
+    }
+
+    /**
+     * Adds a file to the classes layer. Only adds the single source file to the exact path in the
+     * container file system. See {@link LayerConfiguration.Builder#addEntry} for concrete examples
+     * about how the file will be placed in the image.
+     *
+     * @param sourceFile the source file to add to the layer
+     * @param pathInContainer the path in the container file system corresponding to the {@code
+     *     sourceFile}
+     * @return this
+     */
+    public Builder addClassFile(Path sourceFile, AbsoluteUnixPath pathInContainer) {
+      addLayerFile(LayerType.CLASSES, sourceFile, pathInContainer, false);
+      return this;
+    }
+
+    /**
+     * Adds a file to the extra files layer. Only adds the single source file to the exact path in
+     * the container file system. See {@link LayerConfiguration.Builder#addEntry} for concrete
+     * examples about how the file will be placed in the image.
+     *
+     * @param sourceFile the source file to add to the layer
+     * @param pathInContainer the path in the container file system corresponding to the {@code
+     *     sourceFile}
+     * @return this
+     */
+    public Builder addExtraFile(Path sourceFile, AbsoluteUnixPath pathInContainer) {
+      addLayerFile(LayerType.EXTRA_FILES, sourceFile, pathInContainer, false);
+      return this;
+    }
+
+    /**
+     * Adds a file to the exploded WAR layer. Only adds the single source file to the exact path in
+     * the container file system. See {@link LayerConfiguration.Builder#addEntry} for concrete
+     * examples about how the file will be placed in the image.
+     *
+     * @param sourceFile the source file to add to the layer
+     * @param pathInContainer the path in the container file system corresponding to the {@code
+     *     sourceFile}
+     * @return this
+     */
+    // TODO: remove this and put files in WAR into the relevant layers (i.e., dependencies, snapshot
+    // dependencies, resources, and classes layers).
+    public Builder addExplodedWarFile(Path sourceFile, AbsoluteUnixPath pathInContainer) {
+      addLayerFile(LayerType.EXPLODED_WAR, sourceFile, pathInContainer, false);
+      return this;
+    }
+
+    /**
+     * Adds a file to the dependency layer. If the source file is a directory, the directory and its
+     * contents will be added recursively. See {@link LayerConfiguration.Builder#addEntryRecursive}
+     * for concrete examples about how the file will be placed in the image.
+     *
+     * @param sourceFile the source file to add to the layer
+     * @param pathInContainer the path in the container file system corresponding to the {@code
+     *     sourceFile}
+     * @return this
+     */
+    public Builder addDependencyFileRecursive(Path sourceFile, AbsoluteUnixPath pathInContainer) {
```

Comment:
Is it okay if all files are added recursively? (aka are the non-rescursive `add...File` methods necessary?)

Reply:
I tried to be in line with `LayerConfiguration` having both recursive and non-recursive methods, but now I think it is possible to cover all cases only with recursive adds. Then perhaps I can rename `add...FileRecursive()` to just `add...File`.

Reply:
Sounds good!

Reply:
Now I remember. The non-recursive case is useful when the caller is walking a directory, enumerating all files and directories and adding each path.